src/universal/components/SignInPage/SignInPage.js

```diff
@@ -0,0 +1,222 @@
+};
+
+type Props = {
+  atmosphere: Object,
+  dispatch: Dispatch<*>,
+  hasSession: boolean,
+  history: RouterHistory,
+  location: Location,
+  webAuth: Object
+};
+
+type State = {
+  loggingIn: boolean,
+  error: ?Error
+};
+
+const containerStyles = {
+  display: 'flex',
+  flexDirection: 'column'
+};
+
+const loadingErrorStyles = ({
+  display: 'flex',
+  flexDirection: 'column',
+  alignItems: 'center',
+  height: '100%'
+});
+
+// A naive implementation taking a querystring-formatted string (beginning in '?'),
+// and returning an object of the key/value pairs.  Will not work for null-valued
+// keys, but works for our own `?redirectTo=/path/to/page` contract.
+const parseSearch = (search: string): Object => {
+  if (!search) { return {}; }
+  const assocArray = chunkArray(search.slice(1).split(/=|&/), 2);
+  return assocArray.reduce((acc, [key, val]) => ({...acc, [key]: val}), {});
+};
+
+// Note that auth0-js seems to break SSR.  Therefore we dynamically import it
+// over the network.
+const LoadableSignInPage = Loadable({
+  loader: () => import('auth0-js'),
+  loading: LoadableLoading,
+  render: (auth0, props) => {
+    const webAuth = new auth0.WebAuth({
+      domain: __AUTH0_DOMAIN__,
+      clientID: __AUTH0_CLIENT_ID__,
+      redirectUri: window.location.href,
+      scope: 'openid rol tms bet'
+    });
+    return <SignInPage webAuth={webAuth} {...props} />;
+  }
+});
+
+class SignInPage extends Component<Props, State> {
+  state = {
+    error: null,
+    loggingIn: false
+  };
+
+  componentWillMount() {
+    this.maybeRedirectToApp();
+    this.maybeCaptureAuthResponse();
+  }
+
+  componentDidUpdate() {
+    this.maybeRedirectToApp();
+  }
+
+  getHandlerForThirdPartyAuth = (auth0Connection: string) => () => {
+    this.props.webAuth.authorize({
+      connection: auth0Connection,
+      responseType: 'token'
+    });
+  };
+
+  maybeRedirectToApp = () => {
+    if (this.props.hasSession) {
+      const parsedSearch = parseSearch(this.props.location.search);
+      const pathToVisit = parsedSearch.redirectTo || '/me';
+      this.props.history.replace(pathToVisit);
+    }
+  };
+
+  maybeCaptureAuthResponse = () => {
+    // If we've received an auth response, log us in
+    const {hash} = this.props.location;
+    if (hash) {
+      this.setState({loggingIn: true});
+      this.parseAuthResponse(hash)
+        .then(this.saveTokens)
+        .catch((error) => {
+          this.setState({error});
+          throw error; // Inform raven
+        });
+    }
+  };
+
+  parseAuthResponse = (hash: string): Promise<ParsedAuthResponse> => {
+    return new Promise((resolve, reject) => {
+      this.props.webAuth.parseHash({hash}, (err, authResult) => {
+        if (err) {
+          return reject(err);
+        }
+        return resolve(authResult);
+      });
+    });
+  };
+
+  resetState = () => {
+    this.setState({loggingIn: false, error: null});
+  };
+
+  saveTokens = (response: ParsedAuthResponse): void => {
```

Comment:
why plural? it should be a single token (encoded & object form), right?

Reply:
Ah yeah, good catch. Was initially plural because the OIDC response includes an access token as well, but we're not using that.